### PR TITLE
CFY-4936 added maintenance mode

### PR DESCRIPTION
**Pull request by nirbir**

### CODE REVIEW

File: cloudify_cli/commands/executions.py [hunk 1]
```diff
@@ -83,7 +83,7 @@ def ls(deployment_id, include_system_workflows):
             deployment_id=deployment_id,
             include_system_workflows=include_system_workflows)
     except exceptions.CloudifyClientError, e:
-        if not e.status_code != 404:
+        if not e.status_code != 404 or e.status_code == 503:
```
Comment (nirbir) [inline cloudify_cli/commands/executions.py]: remove 503 and test again

File: cloudify_cli/tests/commands/test_status.py [hunk 2]
```diff
@@ -31,9 +31,11 @@
 class StatusTest(CliCommandTest):
 
     def test_status_command(self):
-        self._create_cosmo_wd_settings()
-        self.client.manager.get_status = MagicMock()
-        cli_runner.run_cli('cfy status')
+        with patch('cloudify_rest_client.maintenance.'
+                   'MaintenanceModeClient.status'):
```
Comment (nirbir) [inline cloudify_cli/tests/commands/test_status.py]: make consistenet

File: cloudify_cli/commands/maintenance.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+Handles all commands that start with 'cfy maintenance-mode'
+"""
+
+import time
+
+from cloudify_cli import utils
+from cloudify_cli.logger import get_logger
+from cloudify_cli.cli import get_global_verbosity
+from cloudify_cli.cli import NO_VERBOSE
+from cloudify_cli import exceptions
```
Comment (nirbir) [inline cloudify_cli/commands/maintenance.py]: move after utils

File: cloudify_cli/commands/maintenance.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    logger = get_logger()
+
+    status_response = client.maintenance_mode.status()
+
+    logger.info('\nMaintenance Mode Status:')
+    for param_name, param_value in utils.decode_dict(
+            status_response).iteritems():
+        if param_value and param_name != 'remaining_executions':
+            logger.info('\t{0}: \t{1}'.format(
+                    param_name.replace("_", " "),
```
Comment (nirbir) [inline cloudify_cli/commands/maintenance.py]: title

File: cloudify_cli/commands/maintenance.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    remaining_executions = status_response.remaining_executions
+    if remaining_executions:
+        if len(remaining_executions) == 1:
+            logger.info(
+                    'There is 1 running executions. ')
+        else:
+            logger.info(
+                    'There are {0} running executions. '.format(
+                            len(remaining_executions)))
+        logger.info('Cannot activate maintenance mode yet.')
```
Comment (nirbir) [inline cloudify_cli/commands/maintenance.py]: ?

File: cloudify_cli/commands/maintenance.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+        if get_global_verbosity() != NO_VERBOSE:
+            pt = utils.table(['id', 'deployment_id', 'workflow_id'],
```
Comment (nirbir) [inline cloudify_cli/commands/maintenance.py]: also status

File: cloudify_cli/commands/maintenance.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    if timeout and not wait:
+        msg = "'--timeout' was used without '--wait'."
+        error = exceptions.CloudifyCliError(msg)
+        error.possible_solutions = [
+            "Add the '--wait' flag to the command in order to wait."
+        ]
+        raise error
+
+    logger.info('Activating maintenance mode...')
+    client.maintenance_mode.activate()
```
Comment (nirbir) [inline cloudify_cli/commands/maintenance.py]: talk about it. return activated and not activating

File: cloudify_cli/commands/maintenance.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+    if wait:
+        logger.info("Maintenance mode will be activated once there are "
+                    "no running executions...\n")
+        deadline = time.time() + timeout
+
+        while True:
+            if _is_timeout(timeout, deadline):
+                client.maintenance_mode.deactivate()
```
Comment (nirbir) [inline cloudify_cli/commands/maintenance.py]: talk about it

File: cloudify_cli/commands/maintenance.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+                raise RuntimeError(
+                    'Activating maintenance mode timed out.')
```
Comment (nirbir) [inline cloudify_cli/commands/maintenance.py]: pending above comment

File: cloudify_cli/commands/maintenance.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+    if wait:
+        logger.info("Maintenance mode will be activated once there are "
+                    "no running executions...\n")
+        deadline = time.time() + timeout
+
+        while True:
+            if _is_timeout(timeout, deadline):
+                client.maintenance_mode.deactivate()
+                raise RuntimeError(
```
Comment (nirbir) [inline cloudify_cli/commands/maintenance.py]: cloudifyclierror

File: cloudify_cli/commands/maintenance.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        logger.info('Cannot activate maintenance mode yet.')
+
+        if get_global_verbosity() != NO_VERBOSE:
+            pt = utils.table(['id', 'deployment_id', 'workflow_id'],
+                             remaining_executions)
+            pt.max_width = 50
+            utils.print_table('Remaining executions:', pt)
+
+    if status_response.status == MAINTENANCE_MODE_ACTIVE:
+        logger.info('Info: maintenance mode is active; '
```
Comment (nirbir) [inline cloudify_cli/commands/maintenance.py]: no info

File: cloudify_cli/tests/commands/test_maintenance_mode.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        self.client.maintenance_mode.status = MagicMock()
+        cli_runner.run_cli('cfy maintenance-mode status')
+
+    def test_activate_maintenance(self):
+
+        self.client.maintenance_mode.activate = MagicMock()
+        cli_runner.run_cli('cfy maintenance-mode activate')
+
+    def test_activate_maintenance_with_wait(self):
+
```
Comment (nirbir) [inline cloudify_cli/tests/commands/test_maintenance_mode.py]: no enters bitch

File: cloudify_cli/commands/executions.py [hunk 13]
```diff
@@ -83,7 +83,7 @@ def ls(deployment_id, include_system_workflows):
             deployment_id=deployment_id,
             include_system_workflows=include_system_workflows)
     except exceptions.CloudifyClientError, e:
-        if not e.status_code != 404 or e.status_code == 503:
+        if e.status_code == 404 or e.status_code == 503:
```
Comment (dankilman) [inline cloudify_cli/commands/executions.py]: `if e.status_code != 404`

**Pull request merged**